### PR TITLE
LGBM integration with new client

### PR DESCRIPTION
**Pull request by shnela**

### CODE REVIEW

File: neptune_lightgbm/impl/__init__.py [hunk 1]
```diff
@@ -79,7 +94,79 @@ def __init__(self, run: 'neptune.Run', base_namespace=''):
         self._run['source_code/integrations/neptune-lightgbm'] = __version__
 
     def __call__(self, env):
-        for name, loss_name, loss_value, _ in env.evaluation_result_list:
-            channel_name = '{}{}_{}'.format(self._base_namespace, name, loss_name)
+        # TODO:
+        # (data_name, eval_name, val, is_higher_better): `lightgbm.basic.Booster.__inner_eval`
+        # ('cv_agg', k, np.mean(v), metric_type[k], np.std(v)): `lightgbm.engine._agg_cv_result`
+        for name, loss_name, loss_value, *_ in env.evaluation_result_list:
```
Comment (shnela) [inline neptune_lightgbm/impl/__init__.py]: Here we're interested in only three first values of every entry of `evaluation_result_list` which is generated differently for different `NeptuneCallback` use cases:

#### for `train` and `LGBMClassifier`

It uses [`__inner_eval`](https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/basic.py#L3382) and has 4 values of format: `(data_name, eval_name, val, is_higher_better)`.

#### for `cv`
It uses [`_agg_cv_result`](https://github.com/microsoft/LightGBM/blob/fcd2453514a6ecd780121e4ea31a227559145451/python-package/lightgbm/engine.py#L379) and has 5 values of format: `('cv_agg', k, np.mean(v), metric_type[k], np.std(v))`

Does this code makes sense?
Example results:
* https://app.stage.neptune.ai/wajcha/domain-2-project/e/DOM-977/all?path=foo%2Ffoo2&attribute=training_multi_logloss
* https://app.stage.neptune.ai/wajcha/domain-2-project/e/DOM-978/all?path=foo%2Flgbm_cv%2Ffoo2&attribute=cv_agg_multi_error
* https://app.stage.neptune.ai/wajcha/domain-2-project/e/DOM-979/all?path=foo%2Ffoo2&attribute=validation_multi_error
*

File: neptune_lightgbm/impl/__init__.py [hunk 2]
```diff
@@ -79,7 +94,79 @@ def __init__(self, run: 'neptune.Run', base_namespace=''):
...
+            title="Feature importance (gain)",
+            max_num_features=max_num_features
+        )
+        results_dict["{}feature_importances/split".format(visuals_path)] \
+            = neptune.types.File.as_image(split_plot.figure)
+        results_dict["{}feature_importances/gain".format(visuals_path)] \
+            = neptune.types.File.as_image(gain_plot.figure)
+
+    # TODO: Your file is larger than 15MB. Neptune supports logging files in-memory objects smaller than 15MB.
+    #       Resize or increase compression of this object
```
Comment (shnela) [inline neptune_lightgbm/impl/__init__.py]: For sample code I've tried we've failed to upload some data.

Do we want to fix it somehow, or leave it as it is?

File: neptune_lightgbm/impl/__init__.py [hunk 3]
```diff
@@ -79,7 +94,79 @@ def __init__(self, run: 'neptune.Run', base_namespace=''):
...
+                s = BytesIO()
+                s.write(digraph.pipe(format='png'))
+                s.seek(0)
+                ax.imshow(image.imread(s))
+                ax.axis('off')
+                trees_series.append(neptune.types.File.as_image(ax.figure))
+            results_dict["{}trees".format(visuals_path)] = neptune.types.FileSeries(trees_series)
+        except GraphvizExecutableNotFound:
+            # TODO: link to Neptune docs section where will be such info: https://graphviz.org/download/
+            log.warning(f"Trees won't be logged. To make it work install graphviz library on your OS.")
```
Comment (shnela) [inline neptune_lightgbm/impl/__init__.py]: python `graphviz` is default requirement, but if someone wants to use it, some C-librares are required in operation system.

* Do we  want to display this warning or be more restrictive and raise some kind of error?
* In this message should be link to neptune docs section about graphviz installation.

File: neptune_lightgbm/impl/__init__.py [hunk 4]
```diff
@@ -79,7 +94,79 @@ def __init__(self, run: 'neptune.Run', base_namespace=''):
...
+
+    if log_trees_as_dataframe:
+        if isinstance(booster, lgb.Booster):
+            df = booster.trees_to_dataframe()
+            results_dict["trees_as_dataframe"] = neptune.types.File.as_html(df)
+        else:
+            # TODO: log here and don't perform action, or detect it when calculation starts and raise an exception
+            log.warning("Trees won't be logged as dataframe. `booster` must be instance of `lightgbm.Booster` class.")
```
Comment (shnela) [inline neptune_lightgbm/impl/__init__.py]: `log_trees_as_dataframe` flag can be set only if `booster` of type `Booster` is passed.

Is this action ok, or we should raise exception when user invokes this function with wrong configuration setup.

File: setup.py [hunk 5]
```diff
@@ -15,7 +16,7 @@ def main():
         all_deps += extras[group_name]
     extras['all'] = all_deps
 
-    base_libs = ['neptune-client>=0.5.4', 'lightgbm']
+    base_libs = ['neptune-client>=0.9.0', 'lightgbm', 'matplotlib', 'scikit-plot', 'graphviz']
```
Comment (kamil-kaczmarek) [inline setup.py:19]: @shnela let's make `graphviz` and optional lib, not required dependency. This is because `graphviz` depends on some C libs that user need to install manually, procedure is different for Win/Linux/MacOS.

Check how it's done in XGBoost: https://github.com/neptune-ai/neptune-xgboost/blob/master/neptune_xgboost/impl/__init__.py#L54:L61

Graphsviz is not required: https://github.com/neptune-ai/neptune-xgboost/blob/master/setup.py#L17

File: neptune_lightgbm/impl/__init__.py (hunk 2 above)
Comment (kamil-kaczmarek) [reply neptune_lightgbm/impl/__init__.py]: We want to inform user about:
* fact that upload was unsuccessful,
* what was not logged (tree or model, or sth. else)
* inform that it's a warning not error.

File: neptune_lightgbm/impl/__init__.py (hunk 3 above)
Comment (kamil-kaczmarek) [reply neptune_lightgbm/impl/__init__.py]: UX should be similar to the XGBoost:
* Inform user that she needs to install additional libs
* warning, not error

https://github.com/neptune-ai/neptune-xgboost/blob/master/neptune_xgboost/impl/__init__.py#L54:L61

File: neptune_lightgbm/impl/__init__.py (hunk 4 above)
Comment (kamil-kaczmarek) [reply neptune_lightgbm/impl/__init__.py]: * Your implementation looks good.
* For larger models 15MB limit can happen. (discussed somewhere above.)

File: neptune_lightgbm/impl/__init__.py [hunk 6]
```diff
@@ -79,7 +101,79 @@ def __init__(self, run: 'neptune.Run', base_namespace=''):
         self._run['source_code/integrations/neptune-lightgbm'] = __version__
 
     def __call__(self, env):
```
Comment (kamil-kaczmarek) [inline neptune_lightgbm/impl/__init__.py:107]: It seems that you dropped some features: logging params, and logging feature names. Can you add them as well?

https://github.com/neptune-ai/neptune-client/blob/kk/lgbm/integrations/lightgbm/neptune_lightgbm/impl/__init__.py#L93:L115

File: neptune_lightgbm/impl/__init__.py (hunk 1 above)
Comment (kamil-kaczmarek) [reply neptune_lightgbm/impl/__init__.py]: * Let's refactor: `channel_name` -> `log_name`.
* For **train**, indeed, we need first three values. However, your structure is incorrect -> pls check here: https://github.com/neptune-ai/neptune-client/blob/kk/lgbm/integrations/lightgbm/neptune_lightgbm/impl/__init__.py#L117:L130
(`log_name = "{}{}/{}".format(self._base_namespace, dataset, metric)`)

* for **CV**, we need first three and last values, as we need to log mean and std. Can you adapt it to handle this as well? Pls consult draft implementation for the run structure: https://github.com/neptune-ai/neptune-client/blob/kk/lgbm/integrations/lightgbm/neptune_lightgbm/impl/__init__.py#L117:L130
(`log_val_name = "{}{}/{}/val".format(self._base_namespace, dataset, metric)`)

**Review by kamil-kaczmarek — CHANGES_REQUESTED** [review]
For further runs, please log them directly to this public project: https://app.neptune.ai/common/lightgbm-integration/experiments?split=tbl&dash=charts&viewId=standard-view

**Review by PiotrJander — APPROVED** [review]

File: neptune_lightgbm/impl/__init__.py (hunk 1 above)
Comment (shnela) [reply neptune_lightgbm/impl/__init__.py]: Oh, I've missed that. sorry.

File: neptune_lightgbm/impl/__init__.py (hunk 3 above)
Comment (shnela) [reply neptune_lightgbm/impl/__init__.py]: Ok, let it be this way.

File: neptune_lightgbm/impl/__init__.py (hunk 4 above)
Comment (shnela) [reply neptune_lightgbm/impl/__init__.py]: ok

File: setup.py (hunk 5 above)
Comment (shnela) [reply setup.py:19]: But it's not a problem for user to have  `graphviz` python library installed.

What we're checking when `log_trees` is set, is if user has `graphviz` C libs installed in system.

If we won't require `graphviz` python librarty in `setup.py` user will have to:
* install `graphviz`  in OS manually (as he has to do now)
* install `graphviz` python library in virtualenv manually (now it's handled when our integretion is installed).

File: neptune_lightgbm/impl/__init__.py (hunk 6 above)
Comment (shnela) [reply neptune_lightgbm/impl/__init__.py:107]: +1

File: setup.py (hunk 5 above)
Comment (kamil-kaczmarek) [reply setup.py:19]: yep, You're right, let's keep it like you proposed.

Comment (shnela) [reply setup.py:19]: So what about Xgboost?

We have described problem in that library, which also requires `graphviz`, right?

File: neptune_lightgbm/impl/__init__.py (hunk 2 above)
Comment (shnela) [reply neptune_lightgbm/impl/__init__.py]: Done:
Additional warning is displayed: `'trees_as_dataframe' wasn't logged. Probably generated dataframe was to large`.
When dataframe is non empty, but no data was uploaded to server.

I can't intercept additional `Your file is larger than 15MB....` message, because it's logged in `neptune-client` code.

File: neptune_lightgbm/impl/__init__.py (hunk 3 above)
Comment (kamil-kaczmarek) [reply neptune_lightgbm/impl/__init__.py]: Looks good, can you add "https://graphviz.org" to the warning? We will point user to the correct resource.

File: neptune_lightgbm/impl/__init__.py [hunk 7]
```diff
@@ -61,7 +85,7 @@ class NeptuneCallback:
                            )
     Note:
         If you are running a k-fold validation it is a good idea to add the k-fold prefix
-        and pass it to the `NeptuneCallback` constructor::
+        and pass it to the `NeptuneCallback` constructor::zz
```
Comment (kamil-kaczmarek) [inline neptune_lightgbm/impl/__init__.py:88]:
```suggestion
        and pass it to the `NeptuneCallback` constructor::
```

File: neptune_lightgbm/impl/__init__.py (hunk 2 above)
Comment (kamil-kaczmarek) [reply neptune_lightgbm/impl/__init__.py]: Looks good, maybe we can rephrase it a bit:

```
'trees_as_dataframe' wasn't logged.
Generated DataFrame was too large (>15MB).

Try to log it, by first saving the file to disk, by using:

    df = my_booster.trees_to_dataframe()
    df.to_csv("path/to/file.csv")

and then logging it by using:

    run["trees_as_dataframe"].upload("path/to/file.csv").
```
